### ISSUE TITLE
RedisNodeNotFoundException thrown on Redis Cluster restart

### ISSUE THREAD

**Issue by ijmac3** [redisson/redisson#2256]
Hello, I am currently experiencing an issue with a Redis cluster on Redisson 3.11.2 while testing how our application will handle major outages.

I am providing the cluster config with the address of one of the master nodes. On app startup, it successfully finds all of the other nodes on the cluster. However, if I take down the entire cluster and restart it, Redisson will fail to reconnect to the cluster when it comes back online, throwing a `RedisNodeNotFoundException` whenever it tries to contact Redis. (see Actual behavior for stack trace).

The application reconnects successfully when using a single server setup.

I found this similar issue, https://github.com/redisson/redisson/issues/2234 but upgrading to 3.11.2 has not fixed this for me. 

### Expected behavior
Redisson successfully reconnects to the cluster after the cluster restarts.

### Actual behavior
Throws the following error every time the application attempts to communicate with Redis:
```
org.redisson.client.RedisNodeNotFoundException: Node: NodeSource [slot=8973, addr=null, redisClient=null, redirect=null, entry=null] hasn't been discovered yet
	at org.redisson.connection.MasterSlaveConnectionManager.connectionReadOp(MasterSlaveConnectionManager.java:590) ~[redisson-3.11.2.jar:3.11.2]
	at org.redisson.command.CommandAsyncService.getConnection(CommandAsyncService.java:861) ~[redisson-3.11.2.jar:3.11.2]
	at org.redisson.command.CommandAsyncService.async(CommandAsyncService.java:671) ~[redisson-3.11.2.jar:3.11.2]
	at org.redisson.command.CommandAsyncService$6.run(CommandAsyncService.java:766) ~[redisson-3.11.2.jar:3.11.2]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:682) ~[netty-common-4.1.38.Final.jar:4.1.38.Final]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:757) ~[netty-common-4.1.38.Final.jar:4.1.38.Final]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:485) ~[netty-common-4.1.38.Final.jar:4.1.38.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
```
### Steps to reproduce or test case
Create a RedissonClient with the config specified below. Stop the cluster and restart it.

### Redis version
5.0.5

### Redisson version
3.11.2

### Redisson configuration
```java
config.useClusterServers()
    .addNodeAddress("redis://master_node_IP:7002")
    .setPassword("xxxxxxxxx")
    .setConnectTimeout(15000);
```

If I've missed something or you'd like more information let me know.

**Comment by mrniko** [comment]
It takes time for Redis cluster to discover all nodes. You can check it using `CLUSTER NODES` command

**Comment by ijmac3** [comment]
That was my first thought, but that doesn't seem to be the issue here. I've left it running for over an hour without it detecting even one node. 

To provide some more information, if I create a new RedissonClient object after the cluster restarts, it will discover the nodes and work as expected. However, if I try to reuse a RedissonClient created before the cluster went down it will not reconnect. With a single-server config, I am able to reuse the client without an issue.

**Comment by mrniko** [comment]
Could you set logging level to `TRACE` for `org.redisson` and share the output?

**Comment by ijmac3** [comment]
With trace level logging enabled, it seems that when the cluster starts up again, it does create new connections to the nodes, but then still fails to find them when retrieving data from it.

After the cluster restart I can see the following:
```
<09:39:09> [TRACE] [CommandDecoder] reply: +OK
, channel: [id: 0xee192369, L:/127.0.0.1:53040 - R:127.0.0.1/127.0.0.1:7006], command: CommandData [promise=RedissonPromise [promise=ImmediateEventExecutor$ImmediatePromise@3f7e80be(success)], command=(READONLY), params=[], codec=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1075179085 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x396c8c18, L:/127.0.0.1:53036 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1796673319 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0xfb71a8e2, L:/127.0.0.1:53044 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1820103844 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x6eff02a2, L:/127.0.0.1:53026 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@659325669 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x3cf80991, L:/127.0.0.1:53031 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new pubsub connection created: RedisPubSubConnection@1212446663 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x8d5b30f4, L:/127.0.0.1:53038 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@469317599 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0xfa9c79f8, L:/127.0.0.1:53042 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@321411504 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x3b65fa3a, L:/127.0.0.1:53043 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [INFO] [PubSubConnectionPool] 1 connections initialized for 127.0.0.1/127.0.0.1:7006
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1353755290 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x5ed0eacf, L:/127.0.0.1:53030 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1000408324 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x772d74cf, L:/127.0.0.1:53028 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1765050976 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x87ff3acb, L:/127.0.0.1:53039 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@947956527 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0xee192369, L:/127.0.0.1:53040 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@1831936735 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0x186324bd, L:/127.0.0.1:53041 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [TRACE] [CommandDecoder] reply: +OK
, channel: [id: 0xbbc13af5, L:/127.0.0.1:53035 - R:127.0.0.1/127.0.0.1:7006], command: CommandData [promise=RedissonPromise [promise=ImmediateEventExecutor$ImmediatePromise@5e8b9358(success)], command=(READONLY), params=[], codec=null]
<09:39:09> [DEBUG] [ClientConnectionsEntry] new connection created: RedisConnection@329344355 [redisClient=[addr=redis://127.0.0.1:7006], channel=[id: 0xbbc13af5, L:/127.0.0.1:53035 - R:127.0.0.1/127.0.0.1:7006], currentCommand=null]
<09:39:09> [INFO] [SlaveConnectionPool] 24 connections initialized for 127.0.0.1/127.0.0.1:7006
```

However, I have a process in `DataExtractor` which tries to retrieve or add to a `RBlockingQueue` every few seconds. This continues to throw the following every time:

```
<09:39:56> [DEBUG] [ClusterConnectionManager] slot 8973 for activityTask
<09:39:57> [DEBUG] [ClusterConnectionManager] slot 12962 for removeActivity
<09:39:58> [DEBUG] [CommandAsyncService] attempt 1 for command (LLEN) and params [activityTask]
<09:39:59> [DEBUG] [CommandAsyncService] attempt 1 for command (EVAL) and params [local elemNum = math.min(ARGV[1], redis.call('llen', KEYS[1])) - 1;local vals = redis.call('lrange',..., 1, removeActivity, 100]
<09:40:00> [DEBUG] [CommandAsyncService] attempt 2 for command (LLEN) and params [activityTask]
<09:40:01> [DEBUG] [CommandAsyncService] attempt 2 for command (EVAL) and params [local elemNum = math.min(ARGV[1], redis.call('llen', KEYS[1])) - 1;local vals = redis.call('lrange',..., 1, removeActivity, 100]
<09:40:01> [DEBUG] [CommandAsyncService] attempt 3 for command (LLEN) and params [activityTask]
<09:40:02> [DEBUG] [CommandAsyncService] attempt 3 for command (EVAL) and params [local elemNum = math.min(ARGV[1], redis.call('llen', KEYS[1])) - 1;local vals = redis.call('lrange',..., 1, removeActivity, 100]
<09:40:03> [ERROR] [DataExtractor] Failed to index activities
org.redisson.client.RedisNodeNotFoundException: Node: NodeSource [slot=8973, addr=null, redisClient=null, redirect=null, entry=null] hasn't been discovered yet
	at org.redisson.connection.MasterSlaveConnectionManager.connectionReadOp(MasterSlaveConnectionManager.java:590) ~[redisson-3.11.2.jar:3.11.2]
	at org.redisson.command.CommandAsyncService.getConnection(CommandAsyncService.java:861) ~[redisson-3.11.2.jar:3.11.2]
	at org.redisson.command.CommandAsyncService.async(CommandAsyncService.java:671) ~[redisson-3.11.2.jar:3.11.2]
	at org.redisson.command.CommandAsyncService$6.run(CommandAsyncService.java:766) ~[redisson-3.11.2.jar:3.11.2]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:682) ~[netty-common-4.1.38.Final.jar:4.1.38.Final]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:757) ~[netty-common-4.1.38.Final.jar:4.1.38.Final]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:485) ~[netty-common-4.1.38.Final.jar:4.1.38.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
<09:40:03> [DEBUG] [ClusterConnectionManager] slot 8973 for activityTask
<09:40:04> [DEBUG] [ClusterConnectionManager] slot 12962 for removeActivity
<09:40:04> [DEBUG] [CommandAsyncService] attempt 1 for command (LLEN) and params [activityTask]
<09:40:05> [DEBUG] [CommandAsyncService] attempt 1 for command (EVAL) and params [local elemNum = math.min(ARGV[1], redis.call('llen', KEYS[1])) - 1;local vals = redis.call('lrange',..., 1, removeActivity, 100]
<09:40:06> [DEBUG] [CommandAsyncService] attempt 2 for command (LLEN) and params [activityTask]
<09:40:07> [DEBUG] [CommandAsyncService] attempt 2 for command (EVAL) and params [local elemNum = math.min(ARGV[1], redis.call('llen', KEYS[1])) - 1;local vals = redis.call('lrange',..., 1, removeActivity, 100]
<09:40:08> [DEBUG] [CommandAsyncService] attempt 3 for command (LLEN) and params [activityTask]
<09:40:09> [DEBUG] [CommandAsyncService] attempt 3 for command (EVAL) and params [local elemNum = math.min(ARGV[1], redis.call('llen', KEYS[1])) - 1;local vals = redis.call('lrange',..., 1, removeActivity, 100]
```

The same error also occurs when trying to access Redis in other ways such as adding to a `RLocalCachedMap`.

**Comment by mrniko** [comment]
Thanks for the provided output, but it doesn't contain any information about cluster state.

**Comment by ijmac3** [comment]
I've attached a .txt file with some additional logging. Logs start just before the cluster comes back online.
[Redis Cluster Trace logs.txt](https://github.com/redisson/redisson/files/3491892/Redis.Cluster.Trace.logs.txt)

Does that provide any helpful information?

**Comment by mrniko** [comment]
Sorry, but provided log doesn't contain RedisNodeNotFoundException

**Comment by mrniko** [comment]
I'm looking for log output which contains both `cluster nodes state got from x.x.x.x:nnnn:` and RedisNodeNotFoundException. So I could make ensure that this issue is on Redisson side.

**Comment by ijmac3** [comment]
I've managed to get to the bottom of it. The issue wasn't with Redisson, my apologies.

In development, we were running a redis-cluster on docker using this project: https://github.com/Grokzen/docker-redis-cluster, which we've realized is removing the nodes.conf, dump.rdb and appendonly.aof on a restart and so was not persisting the data. After some hacking about, we've got it persisting and our application is now recovering from a cluster restart as expected.

Thanks for the help anyway!
